src/MongrationDotNet/MigrationRunner.cs

```diff
@@ -1,46 +1,72 @@
-﻿using System.Linq;
+﻿using System.Collections.Generic;
 using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
 using MongoDB.Driver;
 
 namespace MongrationDotNet
 {
     public class MigrationRunner : IMigrationRunner
     {
         private readonly IMongoDatabase database;
-        private IMongoCollection<MigrationDetails> migrationDetailsCollection;
+        private readonly ILogger<MigrationRunner> logger;
+        private readonly IEnumerable<IMigration> migrationCollections;
         private MigrationDetails migrationDetails;
+        private IMongoCollection<MigrationDetails> migrationDetailsCollection;
 
-        public MigrationRunner(IMongoDatabase database)
+        public MigrationRunner(IMongoDatabase database, IEnumerable<IMigration> migrationCollections,
+            ILogger<MigrationRunner> logger = null)
         {
             this.database = database;
+            this.migrationCollections = migrationCollections;
+            this.logger = logger;
         }
+
         public async Task Migrate()
         {
             migrationDetailsCollection = database.GetCollection<MigrationDetails>(Constants.MigrationDetailsCollection);
-            var migrationCollections = MigrationLocator.GetAllMigrations().OrderBy(x => x.Type).ThenBy(x => x.Version);
 
             foreach (var migrationCollection in migrationCollections)
             {
-                var latestAppliedMigration = await migrationDetailsCollection.Find(x => x.Type == migrationCollection.Type && x.Version == migrationCollection.Version).FirstOrDefaultAsync();
-                
-                if (latestAppliedMigration != null) continue;
+                logger?.LogInformation(LoggingEvents.MigrationStarted,
+                    "Migration started for type: {type}, version: {version} and description: {description} ",
+                    migrationCollection.Type, migrationCollection.Version.ToString(), migrationCollection.Description);
+                var latestAppliedMigration = await migrationDetailsCollection
+                    .Find(x => x.Type == migrationCollection.Type && x.Version == migrationCollection.Version)
+                    .FirstOrDefaultAsync();
+
+                if (latestAppliedMigration != null)
+                {
+                    logger?.LogInformation(LoggingEvents.MigrationSkipped,
+                        "Migration has already been applied. Skipped migration for type: {type}, version: {version} and description: {description} ",
+                        migrationCollection.Type, migrationCollection.Version.ToString(),
+                        migrationCollection.Description);
+                    continue;
+                }
```

Comment:
We need to make this more transactionally safe in case 2 nodes run at same to avoid a race condition:
* step1: if migrationDetails for version exists AND is complete we skip it
* step 2: if migrationDetails for version exists AND is not complete AND is not Archived we poll the DB until a configurable timeout (5m?) as it is still running on another node. Once polling shows it is completed we can proceed. If it timed out we should archive it and go to next step
* step 3: Now as version has not started (or has been archived) we should try to do the insert for the migrationDetails record. 
* step 4: If the insert succeeds we know we are the only node doing the migration so apply the migration. OR If the insert fails another node has picked it up so we should go back to step 2


This should remove any race conditions and uses the DB as central point of co-ordination